File: MyPcBuild.ApiService/Domain/Models/Product.cs
using MyPcBuild.ApiService.Domain.Models.Spatial;

namespace MyPcBuild.ApiService.Domain.Models;

/// <summary>
/// Base product record with common properties for all product types.
/// </summary>
public abstract record Product(
    Guid Id,
    string Name,
    decimal Price,
    string Manufacturer
)
{
    /// <summary>
    /// Gets the category name for this product type.
    /// </summary>
    public abstract string CategoryName { get; }
};

/// <summary>
/// Marker interface for products with physical dimensions.
/// </summary>
public interface ISpatialProduct
{
    Dimensions Dimensions { get; }
}

/// <summary>
/// Marker interface for products that provide installation slots.
/// </summary>
public interface ISlottedProduct : ISpatialProduct
{
    List<Slot> Slots { get; }
}

/// <summary>
/// Marker interface for products with internal chambers.
/// </summary>
public interface IChamberedProduct : ISpatialProduct
{
    List<Chamber> Chambers { get; }
}

// Concrete product types by category

public record CpuProduct(
    Guid Id,
    string Name,
    decimal Price,
    string Manufacturer,
    // CPU-specific properties
    CpuSocket Socket,
    int Cores,
    int Threads,
    Frequency BaseClock,
    Frequency BoostClock,
    Power TDP,
    bool IntegratedGraphics
) : Product(Id, Name, Price, Manufacturer)
{
    public override string CategoryName => "CPU";
};

public record MotherboardProduct(
    Guid Id,
    string Name,
    decimal Price,
    string Manufacturer,
    Dimensions Dimensions,
    List<Slot> Slots,
    // Motherboard-specific properties
    CpuSocket Socket,
    string Chipset,
    FormFactor FormFactor,
    MemoryType MemoryType,
    StorageCapacity MaxMemory
) : Product(Id, Name, Price, Manufacturer), ISlottedProduct
{
    public override string CategoryName => "Motherboard";
};

public record GpuProduct(
    Guid Id,
    string Name,
    decimal Price,
    string Manufacturer,
    Dimensions Dimensions,
    List<Slot> Slots,
    // GPU-specific properties
    string ChipsetManufacturer,
    string Series,
    StorageCapacity VRAM,
    MemoryType MemoryType,
    Frequency CoreClock,
    Frequency BoostClock,
    Power TDP,
    Length Length,
    GpuPowerConnector PowerConnectors,
    bool RayTracing
) : Product(Id, Name, Price, Manufacturer), ISlottedProduct
{
    public override string CategoryName => "GPU";
};

public record RamProduct(
    Guid Id,
    string Name,
    decimal Price,
    string Manufacturer,
    // RAM-specific properties
    MemoryType Type,
    StorageCapacity Capacity,
    string Configuration,
    Frequency Speed,
    string CASLatency,
    Voltage Voltage
) : Product(Id, Name, Price, Manufacturer)
{
    public override string CategoryName => "RAM";
};

public record PcCaseProduct(
    Guid Id,
    string Name,
    decimal Price,
    string Manufacturer,
    Dimensions Dimensions,
    List<Chamber> Chambers,
    // PC Case-specific properties
    string FormFactor,
    string Color,
    string SidePanelWindow
) : Product(Id, Name, Price, Manufacturer), IChamberedProduct
{
    public override string CategoryName => "PCCase";
};

public record PsuProduct(
    Guid Id,
    string Name,
    decimal Price,
    string Manufacturer,
    // PSU-specific properties
    Power Wattage,
    string Efficiency,
    string Modular,
    string FormFactor,
    Length Length,
    int PCIe8Pin
) : Product(Id, Name, Price, Manufacturer)
{
    public override string CategoryName => "PSU";
};

public record StorageProduct(
    Guid Id,
    string Name,
    decimal Price,
    string Manufacturer,
    // Storage-specific properties
    string Type,
    string Interface,
    string StorageFormFactor,
    StorageCapacity Capacity,
    DataSpeed ReadSpeed,
    DataSpeed WriteSpeed
) : Product(Id, Name, Price, Manufacturer)
{
    public override string CategoryName => "Storage";
};

public record CoolerProduct(
    Guid Id,
    string Name,
    decimal Price,
    string Manufacturer,
    Dimensions Dimensions,
    // Cooler-specific properties
    CoolerType CoolerType,
    Length Height,
    Power TDP,
    CpuSocket[] Sockets
) : Product(Id, Name, Price, Manufacturer), ISpatialProduct
{
    public override string CategoryName => "Cooler";
};

File: apps/shadcn-vue-client/src/types/product.ts
/**
 * TypeScript contracts for Product Management
 * These interfaces match the API DTOs from the backend
 */

import type { Dimensions, Vector3, Rotation, Slot, Chamber } from './spatial';

// ========== Enums (as string literals for API compatibility) ==========

export type ProductCategory = 'cpu' | 'gpu' | 'motherboard' | 'ram' | 'storage' | 'powersupply' | 'cooler' | 'case';

export type CpuSocket = 'LGA1700' | 'LGA1200' | 'LGA1151' | 'LGA2066' | 'AM5' | 'AM4' | 'sTRX4' | 'TR4';

export type MemoryType = 'DDR3' | 'DDR4' | 'DDR5' | 'GDDR5' | 'GDDR5X' | 'GDDR6' | 'GDDR6X' | 'HBM2' | 'HBM2E' | 'HBM3';

export type FormFactor = 'ATX' | 'MicroATX' | 'MiniITX' | 'EATX';

export type CoolerType = 'Air' | 'AIO' | 'CustomLoop';

export type GpuPowerConnector = 'Dual8Pin' | 'Triple8Pin' | 'One16Pin';

// ========== Value Objects ==========

export interface Frequency {
  valueInGHz: number;
}

export interface StorageCapacity {
  valueInGB: number;
}

export interface Power {
  valueInWatts: number;
}

export interface Voltage {
  valueInVolts: number;
}

export interface Length {
  valueInMm: number;
}

export interface DataSpeed {
  valueInMBps: number;
}

// Re-export spatial types for convenience
export type { Dimensions, Vector3, Rotation, Slot, Chamber };

// ========== Product Base ==========

export interface ProductBase {
  id: string;
  category: ProductCategory;
  name: string;
  price: number;
  manufacturer: string;
  isDraft: boolean;
  publishedAt: string | null;
}

// ========== CPU Product ==========

export interface CpuProductRequest {
  category: 'cpu';
  name: string;
  price: number;
  manufacturer: string;
  socket: CpuSocket;
  cores: number;
  threads: number;
  baseClock: Frequency;
  boostClock: Frequency;
  tdp: Power;
  integratedGraphics: boolean;
}

export interface CpuProductResponse extends ProductBase {
  category: 'cpu';
  socket: CpuSocket;
  cores: number;
  threads: number;
  baseClock: Frequency;
  boostClock: Frequency;
  tdp: Power;
  integratedGraphics: boolean;
}

// ========== GPU Product ==========

export interface GpuProductRequest {
  category: 'gpu';
  name: string;
  price: number;
  manufacturer: string;
  chipsetManufacturer: string;
  series: string;
  vram: StorageCapacity;
  memoryType: MemoryType;
  coreClock: Frequency;
  boostClock: Frequency;
  tdp: Power;
  length: Length;
  powerConnectors: GpuPowerConnector;
  rayTracing: boolean;
  dimensions: Dimensions;
  slots?: Slot[];
}

export interface GpuProductResponse extends ProductBase {
  category: 'gpu';
  chipsetManufacturer: string;
  series: string;
  vram: StorageCapacity;
  memoryType: MemoryType;
  coreClock: Frequency;
  boostClock: Frequency;
  tdp: Power;
  length: Length;
  powerConnectors: GpuPowerConnector;
  rayTracing: boolean;
  dimensions: Dimensions;
  slots?: Slot[];
}

// ========== Motherboard Product ==========

export interface MotherboardProductRequest {
  category: 'motherboard';
  name: string;
  price: number;
  manufacturer: string;
  socket: CpuSocket;
  chipset: string;
  formFactor: FormFactor;
  memoryType: MemoryType;
  maxMemory: StorageCapacity;
  dimensions: Dimensions;
  slots?: Slot[];
}

export interface MotherboardProductResponse extends ProductBase {
  category: 'motherboard';
  socket: CpuSocket;
  chipset: string;
  formFactor: FormFactor;
  memoryType: MemoryType;
  maxMemory: StorageCapacity;
  dimensions: Dimensions;
  slots?: Slot[];
}

// ========== RAM Product ==========

export interface RamProductRequest {
  category: 'ram';
  name: string;
  price: number;
  manufacturer: string;
  type: MemoryType;
  capacity: StorageCapacity;
  configuration: string;
  speed: Frequency;
  casLatency: string;
  voltage: Voltage;
}

export interface RamProductResponse extends ProductBase {
  category: 'ram';
  type: MemoryType;
  capacity: StorageCapacity;
  configuration: string;
  speed: Frequency;
  casLatency: string;
  voltage: Voltage;
}

// ========== Storage Product ==========

export interface StorageProductRequest {
  category: 'storage';
  name: string;
  price: number;
  manufacturer: string;
  type: string;
  interface: string;
  storageFormFactor: string;
  capacity: StorageCapacity;
  readSpeed: DataSpeed;
  writeSpeed: DataSpeed;
}

export interface StorageProductResponse extends ProductBase {
  category: 'storage';
  type: string;
  interface: string;
  storageFormFactor: string;
  capacity: StorageCapacity;
  readSpeed: DataSpeed;
  writeSpeed: DataSpeed;
}

// ========== PSU Product ==========

export interface PsuProductRequest {
  category: 'powersupply';
  name: string;
  price: number;
  manufacturer: string;
  wattage: Power;
  efficiency: string;
  modular: string;
  formFactor: string;
  length: Length;
  pcie8Pin: number;
}

export interface PsuProductResponse extends ProductBase {
  category: 'powersupply';
  wattage: Power;
  efficiency: string;
  modular: string;
  formFactor: string;
  length: Length;
  pcie8Pin: number;
}

// ========== Cooler Product ==========

export interface CoolerProductRequest {
  category: 'cooler';
  name: string;
  price: number;
  manufacturer: string;
  coolerType: CoolerType;
  height: Length;
  tdp: Power;
  sockets: CpuSocket[];
  dimensions: Dimensions;
}

export interface CoolerProductResponse extends ProductBase {
  category: 'cooler';
  coolerType: CoolerType;
  height: Length;
  tdp: Power;
  sockets: CpuSocket[];
  dimensions: Dimensions;
}

// ========== PC Case Product ==========

export interface PcCaseProductRequest {
  category: 'case';
  name: string;
  price: number;
  manufacturer: string;
  formFactor: string;
  color: string;
  sidePanelWindow: string;
  dimensions: Dimensions;
  chambers?: Chamber[];
}

export interface PcCaseProductResponse extends ProductBase {
  category: 'case';
  formFactor: string;
  color: string;
  sidePanelWindow: string;
  dimensions: Dimensions;
  chambers?: Chamber[];
}

// ========== Union Types ==========

export type ProductRequest =
  | CpuProductRequest
  | GpuProductRequest
  | MotherboardProductRequest
  | RamProductRequest
  | StorageProductRequest
  | PsuProductRequest
  | CoolerProductRequest
  | PcCaseProductRequest;

export type ProductResponse =
  | CpuProductResponse
  | GpuProductResponse
  | MotherboardProductResponse
  | RamProductResponse
  | StorageProductResponse
  | PsuProductResponse
  | CoolerProductResponse
  | PcCaseProductResponse;
